tests/writing.rst

```diff
@@ -15,13 +37,48 @@ top), a table of tests (at the bottom), and a tiny main which sets the
 test harness in motion. It uses the table to figure out what the tests
 are called, and what function to call to run them.
 
+Incidentally, when fixing bugs, it's often better to write the test
+before fixing the bug. Firstly, it's easier to assure yourself that the
+bug is (a) genuine, and (b) fixed, because you see the test go from fail
+to pass (though sometimes you don't get the testcase quite right, so
+this isn't doesn't always work as well as it should). Secondly you're
+more likely to write the test carefully, because once you've fixed
+something there's often a feeling that you should commit it for the good
+of the world, which tends to distract you.
+
+The framework is done for you, so you don't need to worry about that
+much. You are responsible for doing two things:
+
+* writing a minimal test or tests for the feature
+* adding that test to the list of tests to be run
+
+Adding the test is simple. There's a ``test_desc`` array in each file that
+comprises a set of tests (I'll come to that in a minute), and you just
+add another entry. The entry is an array consisting of a name for the
+test and a pointer to the function that is the test. Easy. The procedure
+is even simpler for apitest tests - there you just use ``DEFINE_TESTCASE``
+to define your new testcase, and a script picks it up and makes sure it
+is run.
+
+Look at the bottom of ``tests/stemtest.cc`` for the ``test_desc`` array.
+Now look up about 20 lines to where the test functions are defined. You
+need to write a function like these which will return true or false
+depending on whether it failed or not.
+
+In addition, there are a bunch of macros to help you perform standards
+testing tasks. Things like ``TEST_EQUAL`` are all in
+``tests/harness/testsuite.h``. They're pretty simple to use.
```

Comment:
I think we probably want talk about the macros first, since that's how most testcases report failure (and they're better as they allow reporting a reason for the failure).

Perhaps we ought to consider dropping the `bool` return value entirely and replace `return false;` with a suitable `FAIL_TEST("reason");` instead.  Then we'd not have to `return true;` at the end of virtually every test case. It means we have to handle an exception on any test failures, but not on the common path where all tests pass.

Reply:
I think the documentation needs restructuring so that it goes roughly:

1. the functional vs unit test note
2. tests live in tests/
3. apitest works like this, including macros and test databases and probably an example small test (say `termstats`) which shows how it fits together
4. mention stemtest, unittest, internaltest

All that really needs for this to happen (I think) is to move some of the last paragraphs from the opening section (as left by this PR) into the API tests section, editing them to make them only about apitest; add an example apitest in a subheading that includes most of the "API tests" paragraphs; make "test databases" a subhead of "API tests"; add something for the fourth point (stemtest, unittest, internaltest).

I can stick that on my list of things to do, but I don't think it argues for not merging this PR in the meantime. (Particularly because I don't think it helps enough to warrant my working on this in preference to providing feedback on proposals.)

(Changing from return value to exception seems a good idea to me.)

Reply:
Sure, this is an improvement over the current state (especially as this is just existing text moved around).

Reply:
> Changing from return value to exception seems a good idea to me.

I implemented and pushed that a few days ago.